feat(workflows): Track workflow e2e lag in metrics

odin-posthog opened the pull request:
## Problem

There are reports of workflows having excessive e2e lag, and we currently don't have visibility on this lag to see if it is bad or how bad it is.

## Changes

* Bringing "now", the timestamp created in Capture, into the plugin-service as "captured_at" timestamp
* Creating new summary & gauge for e2e lag
* Tracking them after a workflow finishes

## How did you test this code?

Some automated testing for carrying the captured_at timestamp, manual testing for writing the metrics.

Log from my local testing (not part of the PR):
```
[14:11:33.875] INFO (33228): [MAIN] trackE2eLag
    capturedAt: "2025-12-08T13:10:04.540Z"
    e2eLagMs: 89335
```

Log for destinations
<img width="434" height="56" alt="image" src="https://github.com/user-attachments/assets/4b31af98-8def-4850-b470-18063c7b4314" />

odin-posthog committed:
Carry the "now"/"captured_at" timestamp from Capture to plugin-server

odin-posthog committed:
Intermediate commit

odin-posthog committed:
Track e2e lag for workflow executions in metrics

odin-posthog committed:
Fix captured_at not being on the Invocation in the worker

havenbarnes commented on plugin-server/src/main/ingestion-queues/metrics.ts:
```diff
@@ -49,6 +49,18 @@ export const kafkaConsumerEventRequestPendingMsSummary = new Summary({
+export const workflowE2eLagGauge = new Gauge({
+    name: 'workflow_e2e_lag_by_function_ms',
+    help: 'Time difference in ms between event capture time and workflow finishing time, by team and function',
+    labelNames: ['team_id', 'hog_function_id'],
```
also nit: `hog_function_id` -> `hog_flow_id`

havenbarnes commented on plugin-server/src/cdp/consumers/cdp-cyclotron-worker.consumer.ts:
```diff
@@ -37,17 +38,40 @@ export class CdpCyclotronWorker extends CdpConsumerBase {
+            const capturedAt = (result.invocation as CyclotronJobInvocationHogFunction).state?.globals?.event
+                ?.captured_at
+            if (capturedAt) {
+                const e2eLag = Date.now() - new Date(capturedAt).getTime()
```
There's one really tricky thing here that would make this inaccurate: lag caused from delay/"wait until" steps in workflows need to be ignored.

Sadly this wouldn't be as simple as subtracting delay times in the workflow's actions either since not all of them will have taken effect.

I don't know if it's worth over-engineering this too much, so could be good enough for us to simply not track this metric if the workflow has delays 😬

odin-posthog committed:
Track e2e flow at a better suited location

odin-posthog committed:
Slim down metrics to avoid high cardinality in label-combinations

odin-posthog committed:
Fix captured_at type

odin-posthog committed:
Update snapshots to let them know about captured_at

odin-posthog committed:
Add some testing for ingesting captured_at timestamp

odin-posthog committed:
Remove unnecessary comment

odin-posthog committed:
Fix failing test and make e2e lag unit explicit

odin-posthog committed:
Track only summary and add metrics for destinations

Tracking by any useful labels except kafka partition has too high a cardinality, and tracking the partition would require a bunch of extra work that can be left to a followup

odin-posthog committed:
Rename "now" to "captured_at" at an earlier stage

odin-posthog committed:
Merge branch 'master' into feat/odin-add-hogflow-execution-lag-metric

odin-posthog committed:
Track destination e2e lag in the correct place

odin-posthog committed:
Only track destination e2e lag if it's actually finished

havenbarnes commented on plugin-server/src/ingestion/ingestion-consumer.test.ts:
```diff
@@ -79,6 +79,9 @@ const createKafkaMessage = (event: PipelineEvent): Message => {
                 uuid: Buffer.from(event.uuid || ''),
             },
+            {
+                now: Buffer.from(event.now || ''),
```
same here

havenbarnes commented on plugin-server/src/worker/ingestion/process-event.ts:
```diff
@@ -202,7 +202,8 @@ export class EventsProcessor {
         processPerson: boolean,
-        historicalMigration = false
+        historicalMigration = false,
+        captured_at?: Date
```
_super_ nit- capturedAt vs captured_at to match other args?

odin-posthog commented:
HI @jose-sequeira, I put you on this PR because I'm changing some ingest-y things by adding the "captured_at" attribute. I'd appreciate it if you could do a sanity check on that :)

odin-posthog committed:
Rename some more captured_ats to capturedAt

odin-posthog committed:
Add e2e lag tracking to all destination-types and add automated testing

meikelmosby committed:
update tests

meikelmosby committed:
Merge branch 'master' into feat/odin-add-hogflow-execution-lag-metric
